### PR TITLE
Server 42 QinQ support

### PR DESCRIPTION
**Pull request by nikitamarchenko**
~~**Do not merge before #4158**~~

### CODE REVIEW

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java [hunk 1]
```diff
@@ -213,11 +234,42 @@ public OFFlowMod makeDefaultPortServer42IngressFlowMessage(int server42UpdPortOf
...
                 SERVER_42_INGRESS_DEFAULT_FLOW_PRIORITY_OFFSET)
-                .setCookie(U64.of(buildServer42IngressCookie().getValue()))
                 .setMatch(match);
         return makeServer42IngressFlowMessage(builder, Collections.emptyList());
     }
 
+    /**
+     * Make rule to match server 42 packets by inner VLAN tag and forward in in ISL.
+     */
+    public OFFlowMod makeDoubleServer42IngressFlowMessage() {
```
Comment (surabujin) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java:244]: double vlan?

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java [hunk 2]
```diff
@@ -213,11 +234,42 @@ public OFFlowMod makeDefaultPortServer42IngressFlowMessage(int server42UpdPortOf
...
+        FlowEndpoint endpoint = command.getEndpoint();
+        RoutingMetadata metadata = buildServer42IngressMetadata();
+        OFFlowMod.Builder builder = flowModBuilderFactory
+                .makeBuilder(of, TableId.of(SwitchManager.INGRESS_TABLE_ID),
+                        SERVER_42_INGRESS_DOUBLE_VLAN_FLOW_PRIORITY_OFFSET)
```
Comment (surabujin) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java:249]: This constant ruin all efforts of making priorities clear (at least more clear than original design). The idea was to define some base priority and let code to define positive an negarive offsets for this base. As the result - clear vision of order OF flows will be used.

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressInstallFlowModFactory.java [hunk 3]
```diff
@@ -107,7 +107,10 @@ public IngressInstallFlowModFactory(
         return instructions;
     }
 
-    protected abstract List<OFAction> makeServer42IngressFlowTransformActions(List<Integer> vlanStack);
+    protected List<OFAction> makeServer42IngressFlowTransformActions(List<Integer> vlanStack) {
+        // default implimentation to do not force FlowLoop and OneSwitchFlow factories to override this method
+        return new ArrayList<>();
```
Comment (surabujin) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressInstallFlowModFactory.java:112]: Safer to return empty list in class that do not need implementantion, than do raturn empty list in class that must implement it.

**Review by surabujin — APPROVED** [review]

File: src-java/floodlight-service/floodlight-api/src/main/java/org/openkilda/messaging/command/flow/InstallServer42IngressFlow.java [hunk 4]
```diff
@@ -85,39 +85,22 @@ public InstallServer42IngressFlow(@JsonProperty(TRANSACTION_ID) final UUID trans
...
         setEgressSwitchId(egressSwitchId);
         setServer42MacAddress(server42MacAddress);
     }
 
-    /**
-     * Sets output action on the vlan tag.
-     *
-     * @param outputVlanType action on the vlan tag
-     */
-    public void setOutputVlanType(final OutputVlanType outputVlanType) {
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/floodlight-service/floodlight-api/src/main/java/org/openkilda/messaging/command/flow/InstallServer42IngressFlow.java:109]: good point. Agree that we should add verification

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java (hunk 2 above)
Comment (niksv) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java:249]: But this constant is offset. And makeBuilder will calculate priority by `(basePriority + priorityOffset)`

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java (hunk 1 above)
Comment (niksv) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java:244]: yep

File: src-java/floodlight-service/floodlight-modules/src/test/java/org/openkilda/floodlight/command/flow/ingress/IngressServer42FlowInstallCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+        properties.setProperty(SERVER_42_UPD_PORT_OFFSET_PROPERTY, Integer.toString(SERVER_42_UPD_PORT_OFFSET));
+        PropertiesBasedConfigurationProvider provider = new PropertiesBasedConfigurationProvider(properties);
+
+        KildaCore kildaCoreMock = mock(KildaCore.class);
+        expect(kildaCoreMock.getConfig()).andReturn(provider.getConfiguration(KildaCoreConfig.class)).anyTimes();
+        context.addService(KildaCore.class, kildaCoreMock);
+
+        replay(switchServiceMock, featureDetectorServiceMock, kildaCoreMock);
+        command.setup(context);
+        command.setup(context);
```
Comment (niksv) [inline src-java/floodlight-service/floodlight-modules/src/test/java/org/openkilda/floodlight/command/flow/ingress/IngressServer42FlowInstallCommandTest.java:371]: duplicate string

**Review by niksv — COMMENTED** [review]
I think there are some actions need to be done: 
Mandatory:
1. change base of this PR because it contains code from Ready to Merge PR https://github.com/telstra/open-kilda/pull/4154
2. make all-in-one tests green.

Optional:
split this PR by moving Server 42 part (C++ code and code from `java-src/server42`) to separate PR. These two PRs will be independent. I think you should do it because 1200 lines for PR is too much

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java (hunk 2 above)
Comment (surabujin) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactory.java:249]: Yes and it hides offset from view. You must find it's definition and see / remember value to compare with other rules. My opinion - it must be raw number.

To be more clear we should have somethig like `.makeBuilder(of, tableId, PRIOTITY_STEP * N)` where N is desired step and PRIORITY_STEP is 10 (allows to add in futue rules between existing rules).

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/switchmanager/SwitchManager.java [hunk 6]
```diff
@@ -247,8 +246,11 @@
     public static final int ARP_POST_INGRESS_ONE_SWITCH_PRIORITY = FLOW_PRIORITY;
 
     public static final int SERVER_42_INGRESS_DEFAULT_FLOW_PRIORITY_OFFSET = -10;
+    public static final int SERVER_42_INGRESS_DOUBLE_VLAN_FLOW_PRIORITY_OFFSET = 10;
     public static final int SERVER_42_INGRESS_DEFAULT_FLOW_PRIORITY = FLOW_PRIORITY
             + SERVER_42_INGRESS_DEFAULT_FLOW_PRIORITY_OFFSET;
+    public static final int SERVER_42_INGRESS_DOUBLE_VLAN_FLOW_PRIORITY = FLOW_PRIORITY
```
Comment (niksv) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/switchmanager/SwitchManager.java:252]: this constant is unused

**Comment by nikitamarchenko** [comment]
> I think there are some actions need to be done:
> Mandatory:
> 
> 1. change base of this PR because it contains code from Ready to Merge PR #4154
> 2. make all-in-one tests green.
> 
> Optional:
> split this PR by moving Server 42 part (C++ code and code from `java-src/server42`) to separate PR. These two PRs will be independent. I think you should do it because 1200 lines for PR is too much

@niksv  I don't want move 50 loc of c++ to different pr.

**Comment by niksv** [comment]
@nikitamarchenko  please update cookie doc `cookies.md`

**Review by niksv — APPROVED** [review]

**Pull request merged**